Tests fail to compile because the C++11 feature std::result_of_t was removed in C++20

### Does this problem persist on the current master?

- [X] I have verified the issue on the current master

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

```
/usr/local/bin/ccache /usr/local/llvm17/bin/clang++ -DSEQAN3_HAS_BZIP2=1 -DSEQAN3_HAS_ZLIB=1 -I/usr/ports/biology/seqan/work/seqan3-3.3.0-49-g64a431380/test/include -I/usr/ports/biology/seqan/work/seqan3-3.3.0-49-g64a431380/include -isystem /usr/ports/biology/seqan/work/seqan3-3.3.0-49-g64a431380/submodules/sdsl-lite/include -O2 -pipe -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing  -isystem /usr/local/include -O2 -pipe -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing  -isystem /usr/local/include  -DNDEBUG -pedantic -Wall -Wextra -Werror -std=c++20 -pthread -MD -MT alignment/decorator/CMakeFiles/gap_decorator_test.dir/gap_decorator_test.cpp.o -MF alignment/decorator/CMakeFiles/gap_decorator_test.dir/gap_decorator_test.cpp.o.d -o alignment/decorator/CMakeFiles/gap_decorator_test.dir/gap_decorator_test.cpp.o -c /usr/ports/biology/seqan/work/seqan3-3.3.0-49-g64a431380/test/unit/alignment/decorator/gap_decorator_test.cpp
In file included from /usr/ports/biology/seqan/work/seqan3-3.3.0-49-g64a431380/test/unit/alignment/decorator/gap_decorator_test.cpp:19:
In file included from /usr/ports/biology/seqan/work/seqan3-3.3.0-49-g64a431380/test/unit/alignment/decorator/../../alignment/aligned_sequence_test_template.hpp:19:
In file included from /usr/ports/biology/seqan/work/seqan3-3.3.0-49-g64a431380/include/seqan3/io/sam_file/detail/cigar.hpp:24:
/usr/ports/biology/seqan/work/seqan3-3.3.0-49-g64a431380/include/seqan3/io/views/detail/take_until_view.hpp:57:44: error: no member named 'result_of_t' in namespace 'std'
   57 |     static_assert(std::convertible_to<std::result_of_t<fun_t && (std::ranges::range_reference_t<urng_t>)>, bool>,
      |                                       ~~~~~^
/usr/ports/biology/seqan/work/seqan3-3.3.0-49-g64a431380/include/seqan3/io/views/detail/take_until_view.hpp:57:56: error: 'fun_t' does not refer to a value
   57 |     static_assert(std::convertible_to<std::result_of_t<fun_t && (std::ranges::range_reference_t<urng_t>)>, bool>,
      |                                                        ^
/usr/ports/biology/seqan/work/seqan3-3.3.0-49-g64a431380/include/seqan3/io/views/detail/take_until_view.hpp:50:46: note: declared here
   50 | template <std::ranges::view urng_t, typename fun_t, bool or_throw, bool and_consume>
      |                                              ^
/usr/ports/biology/seqan/work/seqan3-3.3.0-49-g64a431380/include/seqan3/io/views/detail/take_until_view.hpp:57:105: error: expected expression
   57 |     static_assert(std::convertible_to<std::result_of_t<fun_t && (std::ranges::range_reference_t<urng_t>)>, bool>,
      |                                                                                                         ^
3 errors generated.
```




### Expected Behavior

n/a

### Steps To Reproduce

n/a

### Environment

```markdown
- Operating system: FreeBSD 13.2
- SeqAn version: 3.3.0-49--g64a431380
- Compiler: clang-16
```


### Anything else?

This feature was removed in C++20, see [here](https://en.cppreference.com/w/cpp/types/result_of).

--- a/.github/workflows/ci_linux.yml
+++ b/.github/workflows/ci_linux.yml
@@ -31,6 +31,12 @@ jobs:
       fail-fast: true
       matrix:
         include:
+          - name: "clang17 libc++"
+            compiler: "clang-17"
+            build: unit
+            build_type: Release
+            cxx_flags: "-stdlib=libc++"
+
           - name: "clang17"
             compiler: "clang-17"
             build: unit

--- a/include/seqan3/core/debug_stream.hpp
+++ b/include/seqan3/core/debug_stream.hpp
@@ -21,7 +21,10 @@
 //!\cond
 namespace std
 {
+namespace
+{
 extern ostream cerr;
+}
 } // namespace std
 //!\endcond
 

--- a/include/seqan3/core/debug_stream/debug_stream_type.hpp
+++ b/include/seqan3/core/debug_stream/debug_stream_type.hpp
@@ -185,12 +185,17 @@ public:
         stream->unsetf(flag);
     }
 
+// fmtflags is an enum in libstdc++ and an unsigned in libc++
+#ifdef _LIBCPP_VERSION
+    static_assert(std::same_as<fmtflags, unsigned>);
+#else
     //!\copybrief setf()
     debug_stream_type & operator<<(fmtflags const flag)
     {
         setf(flag);
         return *this;
     }
+#endif
     //!\}
 
     /*!\name Format flags (seqan3::fmtflags2)

--- a/include/seqan3/io/sam_file/format_bam.hpp
+++ b/include/seqan3/io/sam_file/format_bam.hpp
@@ -401,8 +401,8 @@ format_bam::read_alignment_record(stream_type & stream,
         ref_id = core.refID; // field::ref_id
     }
 
-    flag = core.flag; // field::flag
-    mapq = core.mapq; // field::mapq
+    flag = core.flag;                       // field::flag
+    mapq = static_cast<uint8_t>(core.mapq); // field::mapq
 
     if (core.pos > -1)         // [[likely]]
         ref_offset = core.pos; // field::ref_offset

--- a/include/seqan3/io/sequence_file/format_embl.hpp
+++ b/include/seqan3/io/sequence_file/format_embl.hpp
@@ -105,11 +105,13 @@ protected:
                               id_type & id,
                               qual_type & SEQAN3_DOXYGEN_ONLY(qualities))
     {
+        // Store current position in buffer
+        // Must happen before constructing the view.
+        // With libc++, tellg invalidates the I/O buffer.
+        position_buffer = stream.tellg();
+
         auto stream_view = detail::istreambuf(stream);
         auto stream_it = std::ranges::begin(stream_view);
-
-        // Store current position in buffer.
-        position_buffer = stream.tellg();
 
         std::string idbuffer;
         std::ranges::copy(stream_view | detail::take_until_or_throw(is_cntrl || is_blank),

--- a/include/seqan3/io/sequence_file/format_embl.hpp
+++ b/include/seqan3/io/sequence_file/format_embl.hpp
@@ -197,12 +197,10 @@ protected:
         }
         else
         {
-            detail::consume(stream_view | detail::take_until(is_end));
+            detail::consume(stream_view | detail::take_until_or_throw(is_end)); // consume until "//"
         }
-        //Jump over // and cntrl
-        ++stream_it;
-        ++stream_it;
-        ++stream_it;
+
+        std::ranges::advance(stream_it, 3u, std::ranges::end(stream_view)); // Skip `//` and potentially '\n'
     }
 
     //!\copydoc sequence_file_output_format::write_sequence_record

--- a/include/seqan3/io/sequence_file/format_fasta.hpp
+++ b/include/seqan3/io/sequence_file/format_fasta.hpp
@@ -117,10 +117,12 @@ protected:
                               id_type & id,
                               qual_type & SEQAN3_DOXYGEN_ONLY(qualities))
     {
-        auto stream_view = detail::istreambuf(stream);
-
         // Store current position in buffer
+        // Must happen before constructing the view.
+        // With libc++, tellg invalidates the I/O buffer.
         position_buffer = stream.tellg();
+
+        auto stream_view = detail::istreambuf(stream);
 
         // ID
         read_id(stream_view, options, id);

--- a/include/seqan3/io/sequence_file/format_fastq.hpp
+++ b/include/seqan3/io/sequence_file/format_fastq.hpp
@@ -109,6 +109,11 @@ protected:
                               id_type & id,
                               qual_type & qualities)
     {
+        // Store current position in buffer
+        // Must happen before constructing the view.
+        // With libc++, tellg invalidates the I/O buffer.
+        position_buffer = stream.tellg();
+
         auto stream_view = detail::istreambuf(stream);
         auto stream_it = std::ranges::begin(stream_view);
 
@@ -117,7 +122,6 @@ protected:
         size_t sequence_size_after = 0;
         if constexpr (!detail::decays_to_ignore_v<seq_type>)
             sequence_size_before = size(sequence);
-        position_buffer = stream.tellg();
 
         /* ID */
         if (*stream_it != '@') // [[unlikely]]

--- a/include/seqan3/io/sequence_file/format_genbank.hpp
+++ b/include/seqan3/io/sequence_file/format_genbank.hpp
@@ -164,8 +164,8 @@ protected:
         {
             constexpr auto is_legal_alph = char_is_valid_for<seq_legal_alph_type>;
             std::ranges::copy(
-                stream_view | std::views::filter(!(is_space || is_digit))
-                    | detail::take_until_or_throw_and_consume(is_end) // consume "//"
+                stream_view | std::views::filter(!(is_space || is_digit)) // ignore whitespace and numbers
+                    | detail::take_until_or_throw(is_end)                 // until //
                     | std::views::transform(
                         [is_legal_alph](char const c) // enforce legal alphabet
                         {
@@ -183,9 +183,10 @@ protected:
         }
         else
         {
-            detail::consume(stream_view | detail::take_until_or_throw_and_consume(is_end)); // consume until "//"
-            ++stream_it;                                                                    // consume "/n"
+            detail::consume(stream_view | detail::take_until_or_throw(is_end)); // consume until "//"
         }
+
+        std::ranges::advance(stream_it, 3u, std::ranges::end(stream_view)); // Skip `//` and potentially '\n'
     }
 
     //!\copydoc sequence_file_output_format::write_sequence_record

--- a/include/seqan3/io/sequence_file/format_genbank.hpp
+++ b/include/seqan3/io/sequence_file/format_genbank.hpp
@@ -106,11 +106,13 @@ protected:
                               id_type & id,
                               qual_type & SEQAN3_DOXYGEN_ONLY(qualities))
     {
+        // Store current position in buffer
+        // Must happen before constructing the view.
+        // With libc++, tellg invalidates the I/O buffer.
+        position_buffer = stream.tellg();
+
         auto stream_view = detail::istreambuf(stream);
         auto stream_it = std::ranges::begin(stream_view);
-
-        // Store current position in buffer.
-        position_buffer = stream.tellg();
 
         if (!(std::ranges::equal(stream_view | detail::take_until_or_throw(is_cntrl || is_blank),
                                  std::string{"LOCUS"})))

--- a/include/seqan3/io/stream/detail/fast_istreambuf_iterator.hpp
+++ b/include/seqan3/io/stream/detail/fast_istreambuf_iterator.hpp
@@ -229,6 +229,7 @@ public:
     reference operator*() const
     {
         assert(stream_buf != nullptr);
+        assert(stream_buf->gptr() != stream_buf->egptr());
         return *stream_buf->gptr();
     }
 

--- a/include/seqan3/search/configuration/parallel.hpp
+++ b/include/seqan3/search/configuration/parallel.hpp
@@ -31,7 +31,7 @@ namespace seqan3::search_cfg
  *
  * \include test/snippet/search/configuration_parallel.cpp
  */
-using parallel =
-    seqan3::detail::parallel_mode<std::integral_constant<detail::search_config_id, detail::search_config_id::parallel>>;
+using parallel = seqan3::detail::parallel_mode<
+    std::integral_constant<seqan3::detail::search_config_id, seqan3::detail::search_config_id::parallel>>;
 
 } // namespace seqan3::search_cfg

--- a/include/seqan3/search/detail/search_configurator.hpp
+++ b/include/seqan3/search/detail/search_configurator.hpp
@@ -244,23 +244,20 @@ algorithm_t search_configurator::configure_hit_strategy(configuration_t const & 
         auto cfg_without_hit = cfg.template remove<search_cfg::hit>();
 
         // Apply the correct static configuration element.
-        return std::visit(multi_invocable{[&](search_cfg::hit_all_best)
-                                          {
-                                              return next_config_step(cfg_without_hit | search_cfg::hit_all_best{});
-                                          },
-                                          [&](search_cfg::hit_single_best)
-                                          {
-                                              return next_config_step(cfg_without_hit | search_cfg::hit_single_best{});
-                                          },
-                                          [&](search_cfg::hit_strata const & strata)
-                                          {
-                                              return next_config_step(cfg_without_hit | strata);
-                                          },
-                                          [&](auto)
-                                          {
-                                              return next_config_step(cfg_without_hit | search_cfg::hit_all{});
-                                          }},
-                          hit_variant);
+        if (std::holds_alternative<search_cfg::hit_all_best>(hit_variant))
+        {
+            return next_config_step(cfg_without_hit | search_cfg::hit_all_best{});
+        }
+        else if (std::holds_alternative<search_cfg::hit_single_best>(hit_variant))
+        {
+            return next_config_step(cfg_without_hit | search_cfg::hit_single_best{});
+        }
+        else if (std::holds_alternative<search_cfg::hit_strata>(hit_variant))
+        {
+            return next_config_step(cfg_without_hit | std::get<search_cfg::hit_strata>(hit_variant));
+        }
+        else
+            return next_config_step(cfg_without_hit | search_cfg::hit_all{});
     }
     else // Already statically configured.
     {

--- a/test/unit/alignment/pairwise/edit_distance/proxy_reference_test.cpp
+++ b/test/unit/alignment/pairwise/edit_distance/proxy_reference_test.cpp
@@ -11,17 +11,23 @@
 
 using seqan3::detail::proxy_reference;
 
-namespace std
+template <typename T>
+struct remove_reference
 {
+    using type = std::remove_reference_t<T>;
+};
+
 template <typename T>
 struct remove_reference<proxy_reference<T>>
 {
     using type = T;
 };
-} // namespace std
+
+template <typename T>
+using remove_reference_t = typename remove_reference<T>::type;
 
 template <typename reference_t>
-constexpr bool is_const_ref_v = std::is_const_v<std::remove_reference_t<reference_t>>;
+constexpr bool is_const_ref_v = std::is_const_v<remove_reference_t<reference_t>>;
 
 template <typename T>
 using reference_test = ::testing::Test;

--- a/test/unit/alphabet/views/char_strictly_to_test.cpp
+++ b/test/unit/alphabet/views/char_strictly_to_test.cpp
@@ -79,5 +79,5 @@ TEST(view_char_strictly_to, exception)
     std::string foo = "ACGPTA";
 
     auto v = foo | seqan3::views::char_strictly_to<seqan3::dna5>;
-    EXPECT_THROW((std::ranges::equal(v, "ACGNTA"_dna5)), seqan3::invalid_char_assignment);
+    EXPECT_THROW(static_cast<void>(std::ranges::equal(v, "ACGNTA"_dna5)), seqan3::invalid_char_assignment);
 }

--- a/test/unit/alphabet/views/validate_char_for_test.cpp
+++ b/test/unit/alphabet/views/validate_char_for_test.cpp
@@ -79,5 +79,5 @@ TEST(view_validate_char_for, exception)
     std::string foo = "ACGPTA";
 
     auto v = foo | seqan3::views::validate_char_for<seqan3::dna5>;
-    EXPECT_THROW((std::ranges::equal(v, "ACGNTA"sv)), seqan3::invalid_char_assignment);
+    EXPECT_THROW(static_cast<void>(std::ranges::equal(v, "ACGNTA"sv)), seqan3::invalid_char_assignment);
 }

--- a/test/unit/core/debug_stream_test.cpp
+++ b/test/unit/core/debug_stream_test.cpp
@@ -286,3 +286,14 @@ TEST(debug_stream_test, byte)
     o.flush();
     EXPECT_EQ(o.str(), "40,244");
 }
+
+TEST(debug_stream_test, integers)
+{
+    std::ostringstream o{};
+    seqan3::debug_stream_type my_stream{o};
+
+    my_stream << uint8_t{1} << ',' << uint16_t{2} << ',' << uint32_t{3} << ',' << uint64_t{4} << ',' << size_t{5} << ','
+              << int8_t{6} << ',' << int16_t{7} << ',' << int32_t{8} << ',' << int64_t{9};
+    o.flush();
+    EXPECT_EQ(o.str(), "1,2,3,4,5,6,7,8,9");
+}

--- a/test/unit/io/sequence_file/sequence_file_format_genbank_test.cpp
+++ b/test/unit/io/sequence_file/sequence_file_format_genbank_test.cpp
@@ -50,8 +50,7 @@ DEFINITION  Homo sapiens mRNA for prepro cortistatin like peptide, complete
 ACCESSION   ID3
 ORIGIN
         1 ACGTTTA
-//
-)"};
+//)"};
 
     std::string illegal_alphabet_character_input{
         R"(LOCUS ID1

--- a/test/unit/test/expect_same_type_test.cpp
+++ b/test/unit/test/expect_same_type_test.cpp
@@ -161,12 +161,19 @@ TEST(tuple, same_type_pass)
 
 TEST(tuple, same_type_fail)
 {
+#ifdef _LIBCPP_VERSION
+    char const * error_message = "Expected equality of these values:\n"
+                                 "  decltype(std::tuple{0, .0f, .0, 0u})\n"
+                                 "    Which is: \"std::__1::tuple<int, float, double, unsigned int>\"\n"
+                                 "  std::tuple<int, float, unsigned, double>\n"
+                                 "    Which is: \"std::__1::tuple<int, float, unsigned int, double>\"";
+#else
     char const * error_message = "Expected equality of these values:\n"
                                  "  decltype(std::tuple{0, .0f, .0, 0u})\n"
                                  "    Which is: \"std::tuple<int, float, double, unsigned int>\"\n"
                                  "  std::tuple<int, float, unsigned, double>\n"
                                  "    Which is: \"std::tuple<int, float, unsigned int, double>\"";
-
+#endif
     auto && expect_result =
         seqan3::test::expect_same_type{}("std::type_identity< decltype(std::tuple{0, .0f, .0, 0u})>{}",
                                          "std::type_identity< std::tuple<int, float, unsigned, double>>{}",

--- a/test/unit/test/sequence_generator_test.cpp
+++ b/test/unit/test/sequence_generator_test.cpp
@@ -21,8 +21,13 @@ TEST(random_sequence_generator, fixed_length)
 
     seqan3::test::random_sequence_generator<seqan3::dna4_vector> random_sequence_generator{3};
 
+#ifdef _LIBCPP_VERSION
+    EXPECT_EQ(random_sequence_generator(random_engine), "GTC"_dna4);
+    EXPECT_EQ(random_sequence_generator(random_engine), "GAG"_dna4);
+#else
     EXPECT_EQ(random_sequence_generator(random_engine), "TAG"_dna4);
     EXPECT_EQ(random_sequence_generator(random_engine), "AGC"_dna4);
+#endif
 }
 
 // generate (single) variable sized sequence, generates sequence with size ± size_variance.
@@ -35,9 +40,15 @@ TEST(random_sequence_generator, variable_length_and_different_random_engines)
 
     seqan3::test::random_sequence_generator<seqan3::dna5_vector> random_sequence_generator{4, 2};
 
+#ifdef _LIBCPP_VERSION
+    EXPECT_EQ(random_sequence_generator(random_engine1), "TCCGTT"_dna5);
+    EXPECT_EQ(random_sequence_generator(random_engine2), "CGTAAG"_dna5);
+    EXPECT_EQ(random_sequence_generator(random_engine3), "GA"_dna5);
+#else
     EXPECT_EQ(random_sequence_generator(random_engine1), "CCAN"_dna5);
     EXPECT_EQ(random_sequence_generator(random_engine2), "NN"_dna5);
     EXPECT_EQ(random_sequence_generator(random_engine3), "AG"_dna5);
+#endif
 }
 
 // generate a collection of variable/fixed sized sequences
@@ -54,7 +65,11 @@ TEST(random_sequence_generator, sequence_collection)
                               return random_sequence_generator(random_engine);
                           });
 
+#ifdef _LIBCPP_VERSION
+    std::vector<seqan3::dna5_vector> resulting_sequences = {"CTACN"_dna5, "AC"_dna5, "CTCGG"_dna5, "GGCANN"_dna5};
+#else
     std::vector<seqan3::dna5_vector> resulting_sequences = {"TA"_dna5, "GANG"_dna5, "TGNTCC"_dna5, "NNGC"_dna5};
+#endif
     EXPECT_EQ(sequences, resulting_sequences);
 }
 
@@ -73,9 +88,15 @@ TEST(random_sequence_generator, sequence_pairs)
             return {random_sequence_generator(random_engine), random_sequence_generator(random_engine)};
         });
 
+#ifdef _LIBCPP_VERSION
+    std::vector<std::tuple<seqan3::dna5_vector, seqan3::dna5_vector>> resulting_pairs{{"CTACN"_dna5, "AC"_dna5},
+                                                                                      {"CTCGG"_dna5, "GGCANN"_dna5},
+                                                                                      {"CACA"_dna5, "CNC"_dna5}};
+#else
     std::vector<std::tuple<seqan3::dna5_vector, seqan3::dna5_vector>> resulting_pairs{{"TA"_dna5, "GANG"_dna5},
                                                                                       {"TGNTCC"_dna5, "NNGC"_dna5},
                                                                                       {"GG"_dna5, "AA"_dna5}};
+#endif
 
     EXPECT_EQ(sequence_pairs, resulting_pairs);
 }

--- a/test/unit/utility/container/aligned_allocator_test.cpp
+++ b/test/unit/utility/container/aligned_allocator_test.cpp
@@ -227,6 +227,9 @@ TEST(aligned_allocator, in_list)
 
 TEST(aligned_allocator, in_map)
 {
+#ifdef _LIBCPP_VERSION
+    GTEST_SKIP() << "std::map not aligned in libc++";
+#else
     constexpr size_t alignment = 16;
     using key_type = char;
     using value_type = int;
@@ -251,4 +254,5 @@ TEST(aligned_allocator, in_map)
     EXPECT_EQ(memory_alignment(&*(++it), alignment), 0u);
     EXPECT_EQ(memory_alignment(&*(++it), alignment), 0u);
     EXPECT_EQ(++it, container.end());
+#endif
 }

--- a/test/unit/utility/detail/type_name_as_string_test.cpp
+++ b/test/unit/utility/detail/type_name_as_string_test.cpp
@@ -43,7 +43,11 @@ class type_inspection : public ::testing::Test
                                           "short*",
                                           "double const* const",
                                           "foo::bar<char> const &",
+#ifdef _LIBCPP_VERSION
+                                          "foo::bar<foo::bar<char, double>>"};
+#else
                                           "foo::bar<foo::bar<char, double> >"};
+#endif
 
 public:
     // Returns the name of the type according to the list of names defined above.

